plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hadoop/HadoopIcebergTableOperations.java

```diff
@@ -97,12 +135,260 @@ protected void commitToExistingTable(TableMetadata base, TableMetadata metadata)
+                throw new TrinoException(ICEBERG_COMMIT_ERROR, String.format("Version %d already exists: %s", nextVersion, finalMetadataFile.toString()));
+            }
+            catalog.getTrinoFileSystem().renameFile(tempMetadataFile, finalMetadataFile);
+        }
+        catch (IOException e) {
+            throw new TrinoException(ICEBERG_COMMIT_ERROR, String.format("Failed to commit changes using rename: %s", finalMetadataFile.toString()), e);
+        }
+
+        writeVersionHint(nextVersion);
+        deleteRemovedMetadataFiles(base, metadata);
+        this.shouldRefresh = true;
+    }
+
+    @Override
+    protected void commitMaterializedViewRefresh(TableMetadata base, TableMetadata metadata)
+    {
+        throw new TrinoException(NOT_SUPPORTED, "commitMaterializedViewRefresh is not supported by " + catalog.getCatalogName());
     }
 
     private Table getTable()
     {
         return catalog.loadTable(session, getSchemaTableName());
     }
+
+    private synchronized void updateVersionAndMetadata(int newVersion, String metadataFile)
+    {
+        // update if the current version is out of date
+        if (version.isEmpty() || version.orElseThrow() != newVersion) {
+            this.version = OptionalInt.of(newVersion);
+            this.currentMetadata =
+                    checkUUID(currentMetadata, TableMetadataParser.read(fileIo, metadataFile));
+        }
+    }
+
+    private static TableMetadata checkUUID(TableMetadata currentMetadata, TableMetadata newMetadata)
+    {
+        String newUUID = newMetadata.uuid();
+        if (currentMetadata != null && currentMetadata.uuid() != null && newUUID != null) {
+            checkState(
+                    newUUID.equals(currentMetadata.uuid()),
+                    "Table UUID does not match: current=%s != refreshed=%s",
+                    currentMetadata.uuid(),
+                    newUUID);
+        }
+        return newMetadata;
+    }
+
+    @Override
+    public TableMetadata current()
+    {
+        if (shouldRefresh) {
+            return this.refresh();
+        }
+        return currentMetadata;
+    }
+
+    @Override
+    public TableMetadata refresh()
+    {
+        int ver = version.orElseGet(this::findVersion);
+        //int ver = version.isPresent() ? version : findVersion();
+        try {
+            Optional<Location> metadataFile = getMetadataFile(ver);
+            if (version.isEmpty() && metadataFile.isEmpty() && ver == 0) {
+                // no v0 metadata means the table doesn't exist yet
+                return null;
+            }
+            else if (metadataFile.isEmpty()) {
+                throw new TrinoException(GENERIC_INTERNAL_ERROR, String.format("Metadata file for version %d is missing", ver));
+            }
+
+            Optional<Location> nextMetadataFile = getMetadataFile(ver + 1);
+            while (nextMetadataFile.isPresent()) {
+                ver += 1;
```

Comment:
We should be using the increment operator here
```suggestion
                ver++;
```